File: AdventOfCode2019/08/Day08.hs
module Main where

import Input (digits)
import Data.List (minimumBy)
import Data.List.Split (chunksOf)
import Data.Ord (comparing)

type Layer = String

formatLayer :: Int -> String -> String
formatLayer _ [] = []
formatLayer w list =
  let
    (x, xs) = splitAt w list
    p = map (\t -> if t == '1' then '#' else ' ') x
   in p ++ "\n" ++ formatLayer w xs

count x = length . filter (== x)

part1 :: Int -> Int -> String -> Int
part1 w h digits = count '1' minLayer * count '2' minLayer
  where
    layersList = chunksOf (w*h) digits
    minLayer = minimumBy (comparing (count '0')) layersList

addLayer :: Layer -> Layer -> Layer
addLayer layer1 layer2 =
  [if i1 == '2' then i2 else i1 | (i1, i2) <- zip layer1 layer2]

part2 :: Int -> Int -> String -> Layer
part2 w h digits = image
  where
    layersList = chunksOf (w * h) digits
    image = foldl1 addLayer layersList

main = do
  print $ part1 25 6 digits
  putStr $ formatLayer 25 $ part2 25 6 digits

File: AdventOfCode2019/08/day08.js
var input = require("./input.js");

const layersCount = (digits, w, h) => digits.length / (w * h);

const countChars = (digits, w, h, layer, char) => {
  let count = 0;
  for (let i = layer * w * h; i < (layer + 1) * w * h; i++) {
    if (digits[i] === char) {
      count++;
    }
  }

  return count;
};

const part1 = (digits, w, h) => {
  const layers = layersCount(digits, w, h);
  let minLayer = 0;
  let minZeroCount = w * h;
  for (let i = 0; i < layers; i++) {
    const count = countChars(digits, w, h, i, "0");
    if (count < minZeroCount) {
      minLayer = i;
      minZeroCount = count;
    }
  }

  const answer = countChars(digits, w, h, minLayer, "1") * countChars(digits, w, h, minLayer, "2");
  return answer;
};

const part2 = (digits, w, h) => {
  const transparent = "2";
  const digitsPerLayer = w * h;

  let out = "";
  for (let j = 0; j < h; j++) {
    for (let i = 0; i < w; i++) {
      let position = j * w + i;
      while (digits[position] === transparent) {
        position += digitsPerLayer;
      }
      out += digits[position] === "0" ? "#" : " ";
    }
    out += "\r\n";
  }

  return out;
};

const width = 25;
const height = 6;
console.log(part1(input.digits, width, height));

console.log(part2(input.digits, width, height));
